File: day20.nim
const
  input = 33_100_000
  firstTarget = input div 10
  secondTarget = input div 11


proc solve(target, limit: int): int =
  var houses = newSeq[int](target)
  for i in 1 .. target:
    for j in 1 .. min(target div i, limit):
      houses[j*i-1] += i
    if houses[i-1] > target:
      return i


echo solve(firstTarget, firstTarget)
echo solve(secondTarget, 50)

File: python/day20.py
def solve(target, limit):
    houses = [1] * target
    for i in range(2, target):
        for j in range(min(target//i, limit)):
            houses[i*j] += i
        if houses[i] >= target:
            return i


INPUT = 33100000
print(solve(INPUT//10, INPUT//10))
print(solve(INPUT//11, 50))
